versioned/storage/common/src/main/java/org/projectnessie/versioned/storage/common/logic/ReferenceLogicNonTxImpl.java

```diff
@@ -0,0 +1,619 @@
+ *       Persist#markReferenceAsDeleted(Reference)}.
+ *   <li>Then the reference deletion is tracked by creating a commit in {@link InternalRef#REF_REFS}
+ *       using a {@link Action#REMOVE} operation. Implementations must only commit, if the reference
+ *       is present, to allow the resume/recovery process described below.
+ *   <li>Finally, the {@link Reference} is purged (row physically deleted) via {@link
+ *       Persist#purgeReference(Reference)}.
+ * </ol>
+ *
+ * <h3>Fetching references</h3>
+ *
+ * Accessing references by their full name is performed via {@link Persist#findReference(String)}
+ * (or {@link Persist#findReferences(String[])}) and follows the resume/recovery process described
+ * below.
+ *
+ * <h3>Listing references</h3>
+ *
+ * Listing/querying references is performed via the tip of {@link InternalRef#REF_REFS} and then
+ * {@link Persist#findReferences(String[])} chunks of references to inquire their tips/HEADs.
+ *
+ * <h3>Non-transactional resume/recovery</h3>
+ *
+ * Transactional databases can safely use the native consistent update mechanisms provided by the
+ * database. But non-transactional databases need some logic, which works as follows whenever a
+ * reference is accessed:
+ *
+ * <ul>
+ *   <li>{@link Persist#findReference(String)} (or {@link Persist#findReferences(String[])}) is used
+ *       to fetch a reference by name.
+ *   <li>If the reference is found and {@link Reference#deleted()} is {@code false}, the reference
+ *       has been found and can be returned.
+ *   <li>If the the {@link Reference#deleted()} flag is {@code true}, it is possible that a previous
+ *       delete-reference operation failed in the middle. The implementation must resume the
+ *       reference-deletion process described above.
+ *   <li>If the reference has not been found via {@link Persist#findReference(String)} (or {@link
+ *       Persist#findReferences(String[])}), check whether the reference name exists in {@link
+ *       InternalRef#REF_REFS} and resume the create-reference operation described above.
+ * </ul>
+ *
+ * <h3>Internal references</h3>
+ *
+ * {@link InternalRef Internal references} are neither visible nor can those be updated via this
+ * {@link ReferenceLogicNonTxImpl}.
+ */
+final class ReferenceLogicNonTxImpl implements ReferenceLogic {
+
+  private final Persist persist;
+
+  ReferenceLogicNonTxImpl(Persist persist) {
+    this.persist = persist;
+  }
+
+  @Override
+  @Nonnull
+  @jakarta.annotation.Nonnull
+  public List<Reference> getReferences(
+      @Nonnull @jakarta.annotation.Nonnull List<String> references) {
+    Reference[] refs = persist.findReferences(references.toArray(new String[0]));
+    List<Reference> r = new ArrayList<>(refs.length);
+
+    Supplier<StoreIndex<CommitOp>> refsIndexSupplier = createRefsIndexSupplier();
+
+    for (int i = 0; i < refs.length; i++) {
+      Reference ref = refs[i];
+
+      ref = maybeRecover(references.get(i), ref, refsIndexSupplier);
```

Comment:
I might have mentioned this before (not sure). WDYT about using a unique ID for each reference and tracking that ID in the index (commit against internal refs). Normal commits on a ref would do CAS keyed by that ID as opposed to ref name. This should be simplier is dealing with same names in different reference "generations" (create/delete cycles), I guess. We would not have to assume that recovery happened before the next ref generation is used.

Reply:
IMO tracking an ID  for a reference makes it overly complicated and also eliminates the ability to do range-queries on reference names, and the recovery logic would "explode" (or become impossible). TL;DR all the benefits that the ref-name index brings would be gone.

I doubt there's a real use case by which users re-use reference names that quickly.

Reply:
Range queries use our internal index (from the commit on `int/refs`), right? They do not rely on the backend database, do they?

My thinking was to allocate a unique ID for a ref, then store it in the database, then commit to `int/refs`. The commit will act as the source of truth for listing the references (getting the set of names). Active recovery of partially created / deleted references will not be necessary, only passive cleanup of failed create attempts. Even in case the cleanup fails, it won't invalidate the state of the repository.

An extra hop through the commit at `inf/refs` will be required for reference lookup at the API level, but caching can probably mitigate that. WDYT?